parsec/parsec_internal.h

```diff
@@ -347,35 +348,35 @@ parsec_dependency_t *parsec_hash_find_deps(const parsec_taskpool_t *tp,
                                            parsec_execution_stream_t *es,
                                            const parsec_task_t* task);
 typedef int (parsec_update_dependency_fn_t)(parsec_taskpool_t *tp,
-                                            const parsec_task_t* restrict task,
```

Comment:
Just curious: why remove the `restrict` here?

Reply:
Because this .h file is pulled in by parsec/ttg.h, and restrict is a keyword that is refused by g++ and/or other C++ compilers, even if the declaration is within an `extern "C" { ... }` block. At least, this is what I remember of this issue.

Reply:
Ahh right, we need the header for `parsec_task_s` (and probably some other things). It's a bit ironic that an external project includes `parsec_internal.h` though ^^

Reply:
Some things will probably need cleaning, but I wouldn't say TTG is 'external' from PaRSEC... There is a shade area here :)

Reply:
I don't think the right approach is to remove this optimization keyword. Instead I would propose we protect restrict in parsec_config or parsec_config_bottom for C++ case, and leave it's usages in the C library as intended originally.

Reply:
The macro trick is now implemented.

Reply:
This undef of the macro will have side effects on users that include `parsec.h`

Reply:
Well, it will have a side effect only if the users are using a C++ compiler...

According to stackoverflow, some C++ compilers do support `restrict` or `__restrict__`... And others don't.
I guess we should have some CMake logic to decide if HAVE_RESTRICT or not, and have a corresponding cmakedefine.

CMake has a `c_restrict` low-level C compiler feature testable via `target_compile_features`, but there is no such thing in the C++ compiler feature set. I could try to compile some code with `restrict` with the C++ compiler to see if it's supported. But the macro we define would need to be conditional to the type of compiler anyway.

Comment below if you think I should implement this solution.

Reply:
Does anyone remember why restrict was added in the first place? We don't use it in other places where tp and tasks are passed, so maybe dropping it here is just ok?

Reply:
This is a function pointer and it cannot be inlined, so as long as the compiler doesn't barf on the incompatible function type signatures I don't see what good the restrict does here.